dev-db/clickhouse/clickhouse-19.13.4.32.ebuild

```diff
@@ -0,0 +1,862 @@
+		USE_INTERNAL_ODBC_LIBRARY=no		USE_ODBC		ENABLE_ODBC"
+	"xxhash			-		dev-libs/xxhash[static-libs] \
+		USE_INTERNAL_XXHASH_LIBRARY=no		USE_XXHASH"
+	"zlib-ng		-		zlib-ng-${ZLIB_NG_COMMIT}.tar.gz \
+		USE_INTERNAL_ZLIB_LIBRARY=yes		USE_INTERNAL_ZLIB_LIBRARY"
+	"zstd			-		>=app-arch/zstd-1.3.4[lz4,static-libs] \
+		USE_INTERNAL_ZSTD_LIBRARY=no		ZSTD_LIBRARY"
+)
+
+CLICKHOUSE_CMAKE_VENDORING=()
+CLICKHOUSE_UNPACK_OR_REMOVE=()
+CLICKHOUSE_CMAKE_USE_ARGUMENTS=()
+CLICKHOUSE_CMAKE_CONFIG_CHECKS=()
+
+CLICKHOUSE_VENDORING_HARDCODED=(
+	USE_INTERNAL_BASE64_LIBRARY=yes
+	USE_INTERNAL_ICU_LIBRARY=no
+	USE_INTERNAL_LIBEDIT_LIBRARY=no
+	USE_INTERNAL_LTDL_LIBRARY=no
+	USE_INTERNAL_POCO_MONGODB=yes
+	USE_INTERNAL_READLINE_LIBRARY=no
+)
+
+DEPEND_GEN=""
+RDEPEND_GEN=""
+_clickhouse_setup_vars() {
+	local line component use_flag src dyn internal check activate
+	for line in "${CLICKHOUSE_DEPS[@]}"; do
+		read -r component use_flag src internal check activate <<< "${line}"
+
+		[[ "${src}" == "-" ]] && src=
+		[[ "${use_flag}" == "-" ]] && use_flag=
+
+		if [[ "${internal}" =~ ^.+=yes$ ]]; then
+			[[ "${src}" ]] && CLICKHOUSE_UNPACK_OR_REMOVE+=( "${component}	${src}	${use_flag}" )
+		else
+			CLICKHOUSE_UNPACK_OR_REMOVE+=( "${component}" )
+			dyn="${src/,static-libs(+)/}"
+			dyn="${dyn/,static-libs/}"
+			dyn="${dyn/static-libs(+)/}"
+			dyn="${dyn/static-libs/}"
+			dyn="${dyn/[]/}"
+
+			if [[ "${use_flag}" && "${src}" ]]; then
+				DEPEND_GEN+="		${use_flag}? ( ${src} )"
+				RDEPEND_GEN+="		${use_flag}? ( ${dyn} )"
+			elif [[ "${src}" ]]; then
+				DEPEND_GEN+="		${src}"
+				RDEPEND_GEN+="		${dyn}"
+			fi
+			DEPEND_GEN+=$'\n'
+			RDEPEND_GEN+=$'\n'
+		fi
+
+		[[ ! " ${CLICKHOUSE_VENDORING_HARDCODED[@]} " =~ " ${internal} " ]] && \
+			CLICKHOUSE_CMAKE_VENDORING+=( "${internal} ${use_flag}" )
+
+		CLICKHOUSE_CMAKE_CONFIG_CHECKS+=( "${check} ${use_flag}" )
+
+		if [[ "${activate}" ]]; then
+			CLICKHOUSE_CMAKE_USE_ARGUMENTS+=( "${activate}	${use_flag}" )
+		fi
+	done
+}
+
+_clickhouse_setup_vars
+DEPEND="${DEPEND_BASE}
+	static? (
+${DEPEND_GEN}	)
+"
+RDEPEND="${RDEPEND_BASE}
+	!static? (
+${RDEPEND_GEN}	)
+"
+
+_clang_fullversion() {
+	local ver="$1"; shift
+	set -- $($(tc-getCPP "${@}") -E -P - <<<"__clang_major__ __clang_minor__ __clang_patchlevel__")
+	eval echo "${ver}"
+}
+
+clang-fullversion() {
+	_clang_fullversion '$1.$2.$3' "$@"
+}
+
+clang-version() {
+	_clang_fullversion '$1.$2' "$@"
+}
+
+clang-major-version() {
+	_clang_fullversion '$1' "$@"
+}
+
+clang-minor-version() {
+	_clang_fullversion '$2' "$@"
+}
+
+clang-micro-version() {
+	_clang_fullversion '$3' "$@"
+}
+
+pkg_setup() {
+	use llvm && llvm_pkg_setup
+}
+
+pkg_pretend() {
+	if ! use static; then
+		ewarn "Non static builds are not supported by upstream."
+	fi
+
+	if use static && use '!static-libs'; then
+		ewarn "Static linking is requested but static libraries linking is disabled. "
+		ewarn "Statically linked shared libraries will be created."
```

Comment:
@SoapGentoo what is this?

Reply:
I'm sorry, is this question really directed to me? Could you please elaborate?

Reply:
Is it possible to avoid building static libraries?
see https://flameeyes.blog/2011/08/29/useless-flag-static-libs/
and https://archives.gentoo.org/gentoo-dev/message/2dada80c2b9c85b0e83e6328428bf8ab

Reply:
ClickHouse works best as a static binary (or binaries). Upstream is all about performance testing and assuring. It is used like this in production by many companies, both in containers and standalone.
So by default, I build fully static version of it, but allow user to tweak it in multiple ways.
Perhaps `static-libs` use flag here should be renamed to something else. Right now combination of `static` and `static-libs` will build more or less the same thing as upstream's defaults. Disabling `static` will force the build with dynamic dependencies to external dependencies. Disabling `static-libs` will reduce some size by moving internal pars of the binaries out and installing them as dynamic libraries also (this mode is considered as hack by upstream).